### PR TITLE
Update python version in API and attribution

### LINKED ISSUE #4940: Upgrade API to Python 3.12
## Current Situation

The API runs Python 3.11. It should run the latest Python.

## Suggested Improvement

Update the API to use 3.12:

- Update `api/pyproject.toml` to use 3.12
- Update the API Dockerfile to use 3.12

## Benefit

Use the latest available Python. Don't fall behind and get hit with surprising migration issues when the next major versions roll around, or when Django et co drop support for 3.11.

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #4940 by @sarayourfriend

## Description

Updates python version to 3.12 in the API and python/openverse-attribution. Some packages also had to be updated to work with the new version of Python.

## Testing Instructions

The CI should pass, and the API should work as usual

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/pyproject.toml [hunk 1]
```diff
@@ -8,36 +8,37 @@ authors = [
...
+    "django-migrations-git-conflicts >=1.0, <2",
+    "django-oauth-toolkit >=3.0.1, <4",
+    "django-redis >=5.4.0, <6",
+    "django-split-settings >=1.2.0, <2",
+    "django-structlog >=8.0.0, <9",
+    "django-tqdm >=1.3.1, <2",
+    "django-uuslug >=2.0.0, <3",
+    "djangorestframework >=3.14.0, <4",
+    "drf-spectacular >=0.27.1, <0.28",
+    "elasticsearch==8.13.0",
```
Comment (AetherUnbound) [inline api/pyproject.toml:29]: It looks like this package and `pook` were the only ones that were changed in this list, and both were pinned. Is there a specific version that should be set as the upper bound instead so these can still receive patch updates?

File: api/pyproject.toml [hunk 2]
```diff
@@ -8,36 +8,37 @@ authors = [
...
+    "limit >=0.2.3, <0.3",
+    "pillow >=10.2.0, <11",
+    "psycopg >=3.1.18, <4",
+    "python-decouple >=3.8, <4",
+    "python-xmp-toolkit >=2.0.2, <3",
+    "sentry-sdk >=1.40.6, <2",
+    "uvicorn[standard] >=0.30, <0.31",
+    "openverse-attribution @ file:///${PROJECT_ROOT}/../packages/python/openverse-attribution",
+    "structlog-sentry>=2.1.0",
+    "pook==2.0.0",
```
Comment (AetherUnbound) [inline api/pyproject.toml:41]: Shouldn't this be in the `dev-dependencies` section? Is it needed for core application code?

**Review by AetherUnbound — COMMENTED** [review]
A few notes before we consider merging this!

**Comment by sarayourfriend** [comment]
I was going to suggest looking at going straight to 3.13 (it is GA since last week), but Django 5.1 is required before that's possible (so #4015).

File: api/pyproject.toml (hunk 2 above)
Comment (obulat) [reply api/pyproject.toml]: Removed in [c9a8fcb](https://github.com/WordPress/openverse/pull/5095/commits/c9a8fcbcdcfbb9696f33787bc1a5c5decab5b93a)

File: api/pyproject.toml (hunk 1 above)
Comment (obulat) [reply api/pyproject.toml]: Thank you for catching these! Initially, pdm automatically updated more packages, but some of the tests with pook were failing, so I reverted most unnecessary changes manually, and some reversals resulted in pinned versions :(

**Comment by obulat** [comment]
> I was going to suggest looking at going straight to 3.13 (it is GA since last week), but Django 5.1 is required before that's possible (so #4015).

Great to know. We'll go step by step here :) Some changes in this PR's initial commits caused test failures due to pook mocks not working as expected, so we might have to debug while updating these packages.

File: api/pyproject.toml (hunk 1 above)
Comment (sarayourfriend) [reply api/pyproject.toml]: @obulat if you noticed any issues with Pook, do you mind opening an issue upstream so I can fix it?

**Review by krysal — APPROVED** [review]
I recreated the image and everything seems to be working correctly 👍 

```
env COMPOSE_PROFILES="api" just up --force-recreate --build
just catalog/up && just api/init
```

**Pull request merged**